File: tools/reference/validate-binary-search-tree.py
def isValidBST(root):
    def walk(node, low, high):
        if not node:
            return True
        if not (low < node.val < high):
            return False
        return walk(node.left, low, node.val) and walk(node.right, node.val, high)

    return walk(root, float('-inf'), float('inf'))

File: tools/reference/validate-binary-search-tree.js
// Exercises the LeetCode-style `class Solution` shape.
class Solution {
  isValidBST(root) {
    const walk = (node, low, high) => {
      if (!node) return true;
      if (node.val <= low || node.val >= high) return false;
      return walk(node.left, low, node.val) && walk(node.right, node.val, high);
    };
    return walk(root, -Infinity, Infinity);
  }
}
